### PR TITLE
[rhoai-3.3] patched CVE-2026-33236 - NLTK 3.9.2 to 3.9.4 (pytorch+llmcompressor)

### PR DESCRIPTION
**Pull request by yash2189**
## Description

1.   Upgraded NLTK from 3.9.2 to 3.9.4 to fix two CVEs CVE-2026-33236 and CVE-2026-33231.
2.   Applied surgical edit to two pylock.toml files - swapped version, sdist URL+hash, and wheel URL+hash for the NLTK package entry.

Fixes:                                                                                                                                                     
  - [RHOAIENG-54743](https://redhat.atlassian.net/browse/RHOAIENG-54743) (CVE-2026-33236 - jupyter image)
  - [RHOAIENG-54742](https://redhat.atlassian.net/browse/RHOAIENG-54742) (CVE-2026-33236 - runtime image)
  - [RHOAIENG-54649](https://redhat.atlassian.net/browse/RHOAIENG-54649) (CVE-2026-33231 - runtime image)                                                                                                          
  - [RHOAIENG-54651](https://redhat.atlassian.net/browse/RHOAIENG-54651) (CVE-2026-33231 - jupyter image)
                                                         

## How Has This Been Tested?

  1. Hash verification - Downloaded NLTK 3.9.4 from PyPI and verified SHA256 hashes match pylock.toml entries exactly                                        
  2. Package installation test - Installed NLTK 3.9.4 in clean venv, confirmed version, dependencies, and basic functionality   
  3. Static tests - Ran make test to validate pylock.toml syntax and consistency with pyproject.toml (all relevant tests passed)                             
  4. Local image build attempt - Attempted jupyter pytorch-llmcompressor build 
  
Self checklist (all need to be checked):
- [x] Ensure that you have run `make test` (`gmake` on macOS) before asking for review
- [ ] Changes to everything except `Dockerfile.konflux` files should be done in `odh/notebooks` and automatically synced to `rhds/notebooks`. For Konflux-specific changes, modify `Dockerfile.konflux` files directly in `rhds/notebooks` as these require special attention in the downstream repository and flow to the upcoming RHOAI release.

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

[RHOAIENG-54743]: https://redhat.atlassian.net/browse/RHOAIENG-54743?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RHOAIENG-54742]: https://redhat.atlassian.net/browse/RHOAIENG-54742?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RHOAIENG-54649]: https://redhat.atlassian.net/browse/RHOAIENG-54649?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RHOAIENG-54651]: https://redhat.atlassian.net/browse/RHOAIENG-54651?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

* **New Features**
  * Added pull-request build manifests to retrigger Konflux/PR builds for PyTorch runtime images with LLMCompressor (two runtime targets), including PR filtering, run limits, and build parameterization.

* **Chores**
  * Updated NLTK dependency to 3.9.4 in Python 3.12 runtime environments.

[Review Change Stack](https://app.coderabbit.ai/change-stack/red-hat-data-services/notebooks/pull/2227)

### CODE REVIEW

**Comment by yash2189** [comment]
I can see both images were built successfully as the konflux job was success
https://github.com/red-hat-data-services/notebooks/pull/2227/checks?check_run_id=75561511125
https://github.com/red-hat-data-services/notebooks/pull/2227/checks?check_run_id=75561510730
cc: @san7ket

**Comment by yash2189** [comment]
Tested the image : Jupyter workbench (pytorch+llmcompressor CUDA py312)    
Attached test steps to verify nltk version in notebook
[test_artifacts_jupyter_workbench.txt](https://github.com/user-attachments/files/27705494/test_artifacts_jupyter_workbench.txt)

**Comment by yash2189** [comment]
Successfully tested both workbench and runtime images for the pytorch+llmcompressor component. NLTK has been upgraded to version 3.9.4 and all functional
  tests pass with no regressions.

  ---
  
  ### Test Environment
  - **RHOAI Version**: 3.3.2
  - **Cluster**: OpenShift with RHOAI Operator
  - **PR Commit**: `bb97f4163f91a4c0069fad992512926f6ad6fc3f`
  - **Testing Date**: 2026-05-13

  ---

  ### Images Tested

  | Image Type | Component | Image Reference |
  |:-----------|:----------|:----------------|
  | **Workbench** | odh-workbench-jupyter-pytorch-llmcompressor-cuda-py312 |
  `quay.io/rhoai/pull-request-pipelines:notebooks-bb97f4163f91a4c0069fad992512926f6ad6fc3f-linux-d160-m8xlarge-amd64` |
  | **Runtime** | odh-pipeline-runtime-pytorch-llmcompressor-cuda-py312 |
  `quay.io/rhoai/pull-request-pipelines:notebooks-bb97f4163f91a4c0069fad992512926f6ad6fc3f-linux-extra-fast-amd64@sha256:8b3c60b5...` |

  ---
  
  ### Testing Methodology

  #### Workbench Image Testing
  1. Created a Notebook resource in test namespace
  2. Patched the deployment with the PR workbench image
  3. Executed NLTK-specific functional tests (tokenization, POS tagging)
  4. Executed PyTorch functional tests from `tests/manual/pytorch-test-notebook.ipynb`

  <details>
  <summary>Click to see workbench testing commands</summary>

  ```bash
  # Create test namespace
  oc create namespace nltk-functional-test
  oc label namespace nltk-functional-test opendatahub.io/dashboard=true

  # Deploy and patch workbench
  oc apply -f workbench.yaml
  oc patch notebook pytorch-llmcompressor-wb -n nltk-functional-test --type='json' -p='[
    {"op": "replace", "path": "/spec/template/spec/containers/0/image",
     "value": "quay.io/rhoai/pull-request-pipelines:notebooks-bb97f4163f91a4c0069fad992512926f6ad6fc3f-linux-d160-m8xlarge-amd64"}
  ]'

  oc delete pod pytorch-llmcompressor-wb-0 -n nltk-functional-test
  oc wait --for=condition=Ready pod/pytorch-llmcompressor-wb-0 -n nltk-functional-test --timeout=900s

  # Verify NLTK version
  oc exec -n nltk-functional-test pytorch-llmcompressor-wb-0 -- python3 -c "
  import nltk; print(f'NLTK: {nltk.__version__}')
  import torch; print(f'PyTorch: {torch.__version__}')
  "

  # Run functional tests
  oc exec -n nltk-functional-test pytorch-llmcompressor-wb-0 -- python3 /tmp/nltk-functional-test.py
  oc exec -n nltk-functional-test pytorch-llmcompressor-wb-0 -- python3 /tmp/pytorch-cpu-test.py
  Runtime Image Testing

  1. Created a Pod with the PR runtime image
  2. Executed NLTK-specific functional tests (tokenization, POS tagging)
  3. Executed PyTorch functional tests from tests/manual/pytorch-test-notebook.ipynb

  # Create runtime pod
  oc create namespace nltk-runtime-test
  oc label namespace nltk-runtime-test opendatahub.io/dashboard=true

  cat > /tmp/runtime-pod.yaml << 'EOF'
  apiVersion: v1
  kind: Pod
  metadata:
    name: nltk-runtime-pod
    namespace: nltk-runtime-test
  spec:
    containers:
    - name: test
      image: quay.io/rhoai/pull-request-pipelines:notebooks-bb97f4163f91a4c0069fad992512926f6ad6fc3f-linux-extra-fast-amd64@sha256:8b3c60b5734a81160c02791b3c
  c406a09a6ddeab4e3f1cc9a997b88674bd2d4f
      command: ["sleep", "600"]
      resources:
        limits: {cpu: "2", memory: 8Gi}
        requests: {cpu: "1", memory: 4Gi}
    restartPolicy: Never
  EOF

  oc apply -f /tmp/runtime-pod.yaml
  oc wait --for=condition=Ready pod/nltk-runtime-pod -n nltk-runtime-test --timeout=300s

  # Verify NLTK version
  oc exec -n nltk-runtime-test nltk-runtime-pod -- python3 -c "
  import nltk; print(f'NLTK: {nltk.__version__}')
  import torch; print(f'PyTorch: {torch.__version__}')
  "

  # Run functional tests
  oc exec -n nltk-runtime-test nltk-runtime-pod -- python3 /tmp/nltk-functional-test.py
  oc exec -n nltk-runtime-test nltk-runtime-pod -- python3 /tmp/pytorch-cpu-test.py
  ---
  Test Results

  ✅ Workbench Image Results

  Version Verification:
  NLTK: 3.9.4 ✅
  PyTorch: 2.7.1+cu128

  NLTK Functional Tests:
  🧪 Running: NLTK Import
  ✅ NLTK imported successfully: 3.9.4
  
  🧪 Running: NLTK Tokenization
  ✅ Tokenization works: 11 tokens from test sentence
     Tokens: ['Hello', 'world', '!', 'This', 'is']...
     
  🧪 Running: NLTK POS Tagging
  ✅ POS tagging works: [('The', 'DT'), ('quick', 'JJ'), ('brown', 'NN'), ('fox', 'NN'), ('jumps', 'NNS')]
  
  🧪 Running: PyTorch Integration
  ✅ PyTorch version: 2.7.1+cu128
  ✅ PyTorch CPU tensor creation works
  
  Test Results: 5 passed, 0 failed ✅

  PyTorch Functional Tests:
  1. Working with data...
  ✅ Shape of X [N, C, H, W]: torch.Size([64, 1, 28, 28])
  ✅ Shape of y: torch.Size([64]) torch.int64 
  
  2. Creating Models...
  ✅ Using cpu device
  ✅ Model created: NeuralNetwork
  
  3. Training model (1 epoch)...
    loss: 2.298291  [   64/60000]
    ...
    Test Accuracy: 44.8%, Avg loss: 2.159171
  ✅ Training completed
  
  4. Saving model...
  ✅ Model saved to /tmp/model.pth
  
  5. Loading model...
  ✅ Model loaded
  ✅ Prediction: "Ankle boot", Actual: "Ankle boot"
  
  ✅ All PyTorch tests passed!
  
  ✅ Runtime Image Results

  Version Verification:
  NLTK: 3.9.4 ✅
  PyTorch: 2.7.1+cu128
  
  NLTK Functional Tests:
  🧪 Running: NLTK Import
  ✅ NLTK imported successfully: 3.9.4
  
  🧪 Running: NLTK Tokenization
  ✅ Tokenization works: 11 tokens from test sentence
     Tokens: ['Hello', 'world', '!', 'This', 'is']...
     
  🧪 Running: NLTK POS Tagging
  ✅ POS tagging works: [('The', 'DT'), ('quick', 'JJ'), ('brown', 'NN'), ('fox', 'NN'), ('jumps', 'NNS')]
  
  🧪 Running: PyTorch Integration
  ✅ PyTorch version: 2.7.1+cu128
  ✅ PyTorch CPU tensor creation works
  
  Test Results: 5 passed, 0 failed ✅

  PyTorch Functional Tests:
  1. Working with data...
  ✅ Shape of X [N, C, H, W]: torch.Size([64, 1, 28, 28])
  ✅ Shape of y: torch.Size([64]) torch.int64 
  
  2. Creating Models...
  ✅ Using cpu device
  ✅ Model created: NeuralNetwork
  
  3. Training model (1 epoch)...
    loss: 2.296507  [   64/60000]
    ...
    Test Accuracy: 53.0%, Avg loss: 2.166449
  ✅ Training completed
  
  4. Saving model...
  ✅ Model saved to /tmp/model.pth
  
  5. Loading model...
  ✅ Model loaded
  ✅ Prediction: "Ankle boot", Actual: "Ankle boot"
  
  ✅ All PyTorch tests passed!

  ---

**Comment by yash2189** [comment]
already addressed in separate PR
